pkg/ccl/streamingccl/streamingest/stream_ingestion_dist_test.go

```diff
@@ -176,31 +178,31 @@ func TestSourceDestMatching(t *testing.T) {
 	}
 
 	// validatePairs tests that src-dst assignments are expected.
-	validatePairs := func(sipSpecs []*execinfrapb.StreamIngestionDataSpec, dstNodes []sql.InstanceLocality,
+	validatePairs := func(t *testing.T,
+		sipSpecs map[base.SQLInstanceID]*execinfrapb.StreamIngestionDataSpec,
 		expected map[pair]struct{}) {
-		for i, spec := range sipSpecs {
-
-			// SIPs are created in the order of the destination node ids
-			dstID := dstNodes[i].GetInstanceID()
+		for dstID, spec := range sipSpecs {
+			require.True(t, len(spec.PartitionSpecs) > 0, "empty node %s included in partition specs", dstID)
 			for srcID := range spec.PartitionSpecs {
 				srcIDNum, err := strconv.Atoi(srcID)
 				require.NoError(t, err)
-				_, ok := expected[pair{srcIDNum, int(dstID)}]
+				expectKey := pair{srcIDNum, int(dstID)}
+				_, ok := expected[expectKey]
 				require.True(t, ok, "Src %s,Dst %d do not match", srcID, dstID)
+				delete(expected, expectKey)
 			}
 		}
+		require.Equal(t, 0, len(expected), "expected matches not included")
 	}
 
 	// validateEvenDistribution tests that source node assignments were evenly
 	// distributed across destination nodes. This function is only called on test
 	// cases without an expected exact src-dst node match.
-	validateEvenDistribution := func(sipSpecs []*execinfrapb.StreamIngestionDataSpec, dstNodes []sql.InstanceLocality) {
-
+	validateEvenDistribution := func(t *testing.T, sipSpecs map[base.SQLInstanceID]*execinfrapb.StreamIngestionDataSpec, dstNodes []sql.InstanceLocality) {
+		require.Equal(t, len(sipSpecs), len(dstNodes))
```

Comment:
unless I'm missing something, I don't think this invariant needs to hold.

For example, in this[ test case](https://github.com/cockroachdb/cockroach/blob/a7c6662215a421df4ce11bbae0b00740541eb010/pkg/ccl/streamingccl/streamingest/stream_ingestion_dist_test.go#L280), imagine we had a random dest node in region y. I think this invariant would break. (Might be a good test case to add :D)

Reply:
I suppose it depends on how we use this function going forward. Do we want this to be "even distribution across all dst nodes" or "even distribution across dst nodes that were assigned at least one piece of work"? We iterate over sipSpecs below so if we don't have this assertion, then I think we would return true even if all the work was assigned to one node.

Reply:
ah yeah good point. I think the "even distribution across all dst nodes" definition makes more sense. Perhaps include in the inline comment: "this function assumes that all destination nodes will get assigned some work"

Also, the test case I point to above is exercised [here](https://github.com/cockroachdb/cockroach/pull/109089/files#diff-8f8a1fa80cec90a3bdba05980296c5e5d33a6cea8a203a806bc7205c9cb9109cL251).